frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,937 @@
+
+			let (_, pays) = T::ProposalProvider::close_proposal(
+				proposal_hash,
+				index,
+				proposal_weight_bound,
+				length_bound,
+			)?;
+			if Pays::No == pays {
+				if let Some(Call::kick_member(ref strike)) =
+					proposal.expect("proposal must be exist; qed").is_sub_type()
+				{
+					let strike = T::Lookup::lookup(strike.clone())?;
+					<KickingMembers<T, I>>::remove(strike);
+				}
+			}
+			Ok(().into())
+		}
+
+		/// IInitialize the founders to the given members.
+		///
+		/// This should be called by the referendum and can only be called once.
+		#[pallet::weight(0)]
+		pub fn init_founders(
+			origin: OriginFor<T>,
+			founders: Vec<T::AccountId>,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			ensure!(
+				!Self::has_member(MemberRole::Founder),
+				Error::<T, I>::FoundersAlreadyInitialized
+			);
+			for founder in &founders {
+				Self::has_identity(founder)?;
+			}
+
+			let mut founders = founders;
+			founders.sort();
+			T::InitializeMembers::initialize_members(&founders);
+			Members::<T, I>::insert(&MemberRole::Founder, founders.clone());
+
+			log::debug!(
+				target: "runtime::alliance",
+				"Initialize alliance founders: {:?}",
+				founders,
+			);
+
+			Self::deposit_event(Event::FoundersInitialized(founders));
+			Ok(().into())
+		}
+
+		/// Set a new IPFS cid to the alliance rule.
+		#[pallet::weight(0)]
+		pub fn set_rule(origin: OriginFor<T>, rule: cid::Cid) -> DispatchResultWithPostInfo {
+			T::SuperMajorityOrigin::ensure_origin(origin)?;
+
+			Rule::<T, I>::put(&rule);
+
+			Self::deposit_event(Event::NewRule(rule));
+			Ok(().into())
+		}
+
+		/// Make a new announcement by a new IPFS cid about the alliance issues.
+		#[pallet::weight(0)]
+		pub fn announce(
+			origin: OriginFor<T>,
+			announcement: cid::Cid,
+		) -> DispatchResultWithPostInfo {
+			T::SuperMajorityOrigin::ensure_origin(origin)?;
+
+			let mut announcements = <Announcements<T, I>>::get();
+			announcements.push(announcement);
+			<Announcements<T, I>>::put(announcements);
+
+			Self::deposit_event(Event::NewAnnouncement(announcement));
+			Ok(().into())
+		}
+
+		/// Submit oneself for candidacy.
+		/// Account must have enough transferable funds in it to pay the candidate deposit.
+		#[pallet::weight(0)]
+		pub fn submit_candidacy(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			ensure!(
+				!Self::is_account_blacklist(&who),
+				Error::<T, I>::AlreadyInBlacklist
+			);
+			ensure!(!Self::is_candidate(&who), Error::<T, I>::AlreadyCandidate);
+			ensure!(!Self::is_member(&who), Error::<T, I>::AlreadyMember);
+			// check user self or parent should has verified identity to reuse display name and website.
+			Self::has_identity(&who)?;
+
+			let deposit = T::CandidateDeposit::get();
+			T::Currency::reserve(&who, deposit)
+				.map_err(|_| Error::<T, I>::InsufficientCandidateFunds)?;
+			<DepositOf<T, I>>::insert(&who, deposit);
+
+			Self::add_candidate(&who)?;
```

Comment:
return error after storage write. should wrap in `#[transactional]` and have some tests

Reply:
What do you mean by `should wrap in #[transactional]`?

The account `who` has been checked whether it's a candidate before being called `add_candidate`, so in theory calling `add_candidate` here won't return an error, maybe I should directly use `expect("something")` to indicate this?

Reply:
https://github.com/paritytech/substrate/blob/dbd21b967dfc98a44a43fdb96c468c21df51916e/frame/utility/src/lib.rs#L256

But if it can never fail, then you can do 

```
let res = Self::add_candidate(&who);
debug_assert!(res.is_ok());
```